### PR TITLE
Statically link VCRuntime on Windows with MSVC toolchain

### LINKED ISSUE #2107: Statically link on Windows
When downloading and running `gleam.exe` on a fresh Windows 10, it crashes with the following error message: _The program can't start because VCRUNTIME140.dll is missing from your computer._

I believe this is because `gleam.exe` is dynamically linked to the MSVC runtime so, in order to run Gleam, users will have to install a Visual C++ Redistributable first (which is not mentioned in the documentation).

### PR DESCRIPTION
**Pull request by abs0luty**
Resolves #2107

### CODE REVIEW

**Review by lpil — COMMENTED** [review]
Thank you.

This ticket was for the release build rather than all builds. It'll need to leave the regular one unmodified and have an example of the release build successfully working in another repo.

**Comment by abs0luty** [comment]
> and have an example of the release build successfully working in another repo.

I didn't get your point. Can I add building Gleam in a release on Windows into CI then?

**Comment by abs0luty** [comment]
@lpil https://github.com/abs0luty/gleam/actions/runs/7223208851

File: compiler-cli/.cargo/config.toml [hunk 1]
```diff
@@ -0,0 +1,4 @@
+# Statically link the CRT on Windows with MSVC toolchain
+# (only in release mode)
+[target.'cfg(all(windows, target_env = "msvc", not(debug_assertions)))']
```
Comment (lpil) [inline compiler-cli/.cargo/config.toml:3]: What's this debug_assertions part?

**Comment by lpil** [comment]
I can't see the compiled binary for that build, where can I download it? It also doesn't seem to actually be running commit [c55cd35](https://github.com/gleam-lang/gleam/pull/2469/commits/c55cd356756952e45c008ab3035f70d5dedcc972)

**Comment by abs0luty** [comment]
@lpil compiled binary can be found in artifacts or [here](https://github.com/abs0luty/gleam/actions/runs/7223208851/artifacts/1117468157).

File: compiler-cli/.cargo/config.toml (hunk 1 above)
Comment (abs0luty) [reply compiler-cli/.cargo/config.toml:3]: Turned by default in `dev` profile.

File: compiler-cli/Cargo.toml [hunk 2]
```diff
@@ -86,3 +86,8 @@ camino = { version = "1.1.6", features = ["serde1"] }
 pretty_assertions = "1.0.0"
 # Creation of temporary directories
 tempfile = "3.2.0"
+
+[build-dependencies]
+# For statically linking the VCRuntime on Windows when
+# using the MSVC toolchain
+static_vcruntime = "2.0.0"
```
Comment (lpil) [inline compiler-cli/Cargo.toml:93]: Can this be included only for release builds + on Windows?

File: compiler-cli/build.rs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+fn main() {
+    static_vcruntime::metabuild();
```
Comment (lpil) [inline compiler-cli/build.rs:2]: Can this be included only for release builds + on Windows?

Comment (lpil) [inline compiler-cli/build.rs:2]: Can this be included only for release builds + on Windows?

Comment (abs0luty) [reply compiler-cli/build.rs:2]: If you will checkout implementation of `metabuild()` function, it early returns on not release builds or builds without msvc:

```rs
pub fn metabuild() {
    if env::var("CARGO_MANIFEST_DIR").is_err() {
        panic!("`metabuild` must be called from a build script");
    }
	println!("cargo:rerun-if-env-changed=PROFILE");

	// Early exit if not msvc or release
	if env::var("CARGO_CFG_TARGET_ENV").as_deref() != Ok("msvc") || env::var("PROFILE").as_deref() != Ok("release") {
		return;
	}
    
    override_msvcrt_lib();

	// Disable conflicting libraries that aren't hard coded by Rust.
	println!("cargo:rustc-link-arg=/NODEFAULTLIB:libvcruntimed.lib");
	println!("cargo:rustc-link-arg=/NODEFAULTLIB:vcruntime.lib");
	println!("cargo:rustc-link-arg=/NODEFAULTLIB:vcruntimed.lib");
	println!("cargo:rustc-link-arg=/NODEFAULTLIB:libcmtd.lib");
	println!("cargo:rustc-link-arg=/NODEFAULTLIB:msvcrt.lib");
	println!("cargo:rustc-link-arg=/NODEFAULTLIB:msvcrtd.lib");
	println!("cargo:rustc-link-arg=/NODEFAULTLIB:libucrt.lib");
	println!("cargo:rustc-link-arg=/NODEFAULTLIB:libucrtd.lib");
    // Set the libraries we want.
	println!("cargo:rustc-link-arg=/DEFAULTLIB:libcmt.lib");
	println!("cargo:rustc-link-arg=/DEFAULTLIB:libvcruntime.lib");
	println!("cargo:rustc-link-arg=/DEFAULTLIB:ucrt.lib");
}
```

**Pull request merged**